### PR TITLE
Multi discrepancies

### PR DESCRIPTION
**Pull request by danielkec**
Signed-off-by: Daniel Kec <daniel.kec@oracle.com>

### CODE REVIEW

**Review by olotenko — CHANGES_REQUESTED** [review]

File: common/reactive/src/main/java/io/helidon/common/reactive/DeferredScalarSubscription.java [hunk 1]
```diff
@@ -31,91 +32,66 @@
...
-                // state == COMPLETE
-                // state == HAS_VALUE_NO_REQUEST
-                // state == HAS_VALUE_HAS_REQUEST
-                // state == CANCELED
-                break;
-            }
+        value = item; // assert: even if the race occurs, we will deliver one of the items with which complete()
+        //         has been invoked - we support only the case with a single invocation of complete()
+        int state = getAndUpdate(n -> n | VALUE_ARRIVED);
+        if ((state & REQUEST_ARRIVED) == REQUEST_ARRIVED) {
```
[the remark this answers is not in the document]
Comment (spericas) [reply common/reactive/src/main/java/io/helidon/common/reactive/DeferredScalarSubscription.java:89]: I agree that the masking in line 89 is redundant.

File: common/reactive/src/main/java/io/helidon/common/reactive/MultiConcatArray.java [hunk 2]
```diff
@@ -98,6 +98,7 @@ public void nextSource() {
         @Override
         public void request(long n) {
             if (n <= 0) {
+                cancel();
```
[the remark this answers is not in the document]
Comment (spericas) [reply common/reactive/src/main/java/io/helidon/common/reactive/MultiConcatArray.java:101]: @olotenko To be clear, your point is that the sequence `onError onNext` is still possible as cancellation is only a best-effort task.

File: common/reactive/src/main/java/io/helidon/common/reactive/DeferredScalarSubscription.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielkec) [reply common/reactive/src/main/java/io/helidon/common/reactive/DeferredScalarSubscription.java:89]: Aha I see, but it makes sense in the other complete() right?
```java
    public final void complete() {
        int state = get();
        if (((state & VALUE_ARRIVED) != VALUE_ARRIVED) && compareAndSet(state, DONE)){
            downstream.onComplete();
        }
    }
```

[the remark this answers is not in the document]
Comment (olotenko) [reply common/reactive/src/main/java/io/helidon/common/reactive/DeferredScalarSubscription.java]: Yes, in the other `complete` it reflects the logic of the POC correctly: basically, in that `complete` it is "if value has not arrived yet" (was worded with different conditions with OR).

But in this `complete` the logic must be "if request has been seen, but the value still not seen" - only in that case we are allowed to deliver the signals to downstream.

File: common/reactive/src/main/java/io/helidon/common/reactive/MultiConcatArray.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (olotenko) [reply common/reactive/src/main/java/io/helidon/common/reactive/MultiConcatArray.java:101]: Yes, you can view this as the best effort so far. But like I mentioned offline, there is another issue that needs addressing, so we may just as well postpone touching this class.

**Review by olotenko — DISMISSED** [review]
I am indifferent to the change in `MultiConcatArray`. `DeferredScalarSubscription` is good.

**Review by spericas — APPROVED** [review]

**Pull request merged**